### PR TITLE
Made the field definition lookup more optimised

### PR DESCRIPTION
**Pull request by bbakerman**
The field lookup code will most of the time be looking up a object type field and not a system field.  So this adds a more optimised code path for that.

As seen in some memory profiling - these assertions are not cost free

```
Recorded Objects
Session:	graphql-java-clean [:ComplexQueryBenchmark.main()]
Time of export:	Saturday, March 2, 2024 at 4:11:45 PM Australian Eastern Daylight Time
JVM time:	01:25
 	 
Sorted by: 	Instance Count
Aggregation level: 	Classes
Liveness mode: 	Live Objects
Recording rate: 	1/10 allocations

Name													Instance Count	Size
java.util.LinkedHashMap$Entry								16,915 (12 %)	676 kB
java.util.concurrent.CompletableFuture						16,644 (12 %)	399 kB
java.lang.String											14,836 (10 %)	356 kB
java.lang.Object[ ]											5,111 (3 %)	135 kB
graphql.introspection.Introspection$$Lambda$592.154913348	 3,877 (2 %)	93,048 bytes
graphql.introspection.Introspection$$Lambda$588.1661450720	 3,855 (2 %)	61,680 bytes
com.google.common.collect.ImmutableList$Builder				3,553 (2 %)	85,272 bytes

```

`graphql.introspection.Introspection$$Lambda$592.154913348` and `graphql.introspection.Introspection$$Lambda$588.1661450720` are the assertion allocated supplier messages

So while small - there is some memory cost to assertions

### CODE REVIEW

File: src/main/java/graphql/introspection/Introspection.java [hunk 1]
```diff
@@ -702,6 +702,43 @@ public static boolean isIntrospectionTypes(GraphQLNamedType type) {
...
+        if (fieldDefinition != null) {
+            return fieldDefinition;
+        }
+
+        assertTrue(parentType instanceof GraphQLFieldsContainer, () -> String.format("should not happen : parent type must be an object or interface %s", parentType));
+        GraphQLFieldsContainer fieldsContainer = (GraphQLFieldsContainer) parentType;
+        fieldDefinition = schema.getCodeRegistry().getFieldVisibility().getFieldDefinition(fieldsContainer, fieldName);
+        assertTrue(fieldDefinition != null, () -> String.format("Unknown field '%s' for type %s", fieldName, fieldsContainer.getName()));
+        return fieldDefinition;
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/introspection/Introspection.java:715]: I have left the old method in place for semantic reasons and also its signature is not what the execution engine needs.  

The `parentType instanceof GraphQLFieldsContainer` is not needed say.

File: src/main/java/graphql/introspection/Introspection.java [hunk 2]
```diff
@@ -702,6 +702,43 @@ public static boolean isIntrospectionTypes(GraphQLNamedType type) {
...
+            // we look up system fields second because they are less likely to be the field in question
+            fieldDefinition = getSystemFieldDef(schema, parentType, fieldName);
+            if (fieldDefinition == null) {
+                Assert.assertShouldNeverHappen(String.format("Unknown field '%s' for type %s", fieldName, parentType.getName()));
+            }
+        }
+        return fieldDefinition;
+    }
+
+    private static GraphQLFieldDefinition getSystemFieldDef(GraphQLSchema schema, GraphQLCompositeType parentType, String fieldName) {
```
Comment (bbakerman) [inline src/main/java/graphql/introspection/Introspection.java:741]: I made a common private method for the system field lookup code to share some between the two methods

File: src/main/java/graphql/introspection/Introspection.java [hunk 3]
```diff
@@ -702,6 +702,43 @@ public static boolean isIntrospectionTypes(GraphQLNamedType type) {
...
+        GraphQLFieldDefinition fieldDefinition = schema.getCodeRegistry().getFieldVisibility().getFieldDefinition(parentType, fieldName);
+        if (fieldDefinition == null) {
+            // we look up system fields second because they are less likely to be the field in question
+            fieldDefinition = getSystemFieldDef(schema, parentType, fieldName);
+            if (fieldDefinition == null) {
+                Assert.assertShouldNeverHappen(String.format("Unknown field '%s' for type %s", fieldName, parentType.getName()));
+            }
+        }
+        return fieldDefinition;
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/introspection/Introspection.java:739]: The comments above outlines the reasoning

**Comment by andimarek** [comment]
deprecate the old method

**Pull request merged**